### PR TITLE
Add option to serialize private members from base types

### LINKED ISSUE #36: Add option to serialize private members from base types
If you serialize a class B that derives from A then the private fields of A are not serialized even if you apply the correct type attributes. The reason is in the `GetFieldsToBeSerialized(...)` method. GetMembers() does not deliver private fields of base types.

`foreach (var member in typeWrapper.UnderlyingType.GetMembers(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public))`

To solve this issue I iterate through the base types as well and I introduced a member wrapper cache which improves the serialization/deserialisation performance.
You find the modified serializer source code attached.
[YAXSerializer.cs.txt](https://github.com/YAXLib/YAXLib/files/49747/YAXSerializer.cs.txt)

Thanks in advance,
Bernhard

### PR DESCRIPTION
**Pull request by axunonb**
Closes #36

* Added option `YAXSerializableTypeAttribute.IncludePrivateMembersFromBaseTypes`
* Added extension method for `Type,GetMembers(Type, BindingFlags, bool),` used in `YAXSerializer.GetFieldsToBeSerialized(UdtWrapper)`
* Added property `UdtWrapper.IncludePrivateMembersFromBaseTypes`
* Added unit tests
* Modfied xmldoc with references to that option
* Fixed xmldoc in `YAXSerializationOptions`

### CODE REVIEW

**Comment by axunonb** [comment]
The suggestion in #36 to **always** include private members from base types would mean a breaking change, and also with several unit tests tailing. The changed behavior would also mean a deviation from dot net `Type.GetMembers(BindingFlags)`. So it's better to introduce an explicit option to include private members from base types when actually required.

File: YAXLib/Attributes/YAXSerializableTypeAttribute.cs [hunk 1]
```diff
@@ -81,6 +82,12 @@ public YAXSerializationOptions Options
             }
         }
 
+        /// <summary>
+        ///     If set to <see langword="true"/>, private fields from <see cref="Type.BaseType"/>s
+        ///     can be searched for fields to serialize. Default is <see langword="false"/>.
+        /// </summary>
+        public bool IncludePrivateMembersFromBaseTypes { get; set; }
```
Comment (304NotModified) [inline YAXLib/Attributes/YAXSerializableTypeAttribute.cs:89]: Maybe this should be a member of YAXLib/Enums/YAXSerializationOptions.cs - it's already a flags enum.

(member of YAXSerializationFields would be better, but that's is a really large refactor)

File: YAXLib/TypeExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        /// Both, <see cref="Type.GetProperty(string, BindingFlags)"/> and <see cref="Type.GetMembers()"/>
+        /// return private properties/members.
+        /// <see cref="Type.GetMembers(BindingFlags)"/>, however, <b>does not</b> include members from <b>base types</b>,
+        /// whereas <see cref="Type.GetProperty(string, BindingFlags)"/> also returns private properties of base types.
+        /// </remarks>
+        public static List<MemberInfo> GetMembers(this Type t, BindingFlags bindingFlags,
+            bool includeBaseTypePrivateMembers)
+        {
+            var memberList = new List<MemberInfo>();
+            memberList.AddRange(t.GetMembers(bindingFlags));
```
Comment (304NotModified) [inline YAXLib/TypeExtensions.cs:43]: maybe we could do `t.GetMembers(bindingFlags)` first, so we could set a better initial capacity

something like this:

```c#
var currentMembers = t.GetMembers(bindingFlags); // Maybe needs a ToList()
var memberList = new List<MemberInfo>(currentMembers .Count * 2);
memberList.AddRange(currentMembers );
```

For performance reasons

File: YAXLib/Attributes/YAXSerializableTypeAttribute.cs (hunk 1 above)
Comment (axunonb) [reply YAXLib/Attributes/YAXSerializableTypeAttribute.cs:89]: Good point, yep

File: YAXLib/TypeExtensions.cs (hunk 2 above)
Comment (axunonb) [reply YAXLib/TypeExtensions.cs:43]: Agree :)

File: YAXLib/Attributes/YAXSerializableTypeAttribute.cs (hunk 1 above)
Comment (axunonb) [reply YAXLib/Attributes/YAXSerializableTypeAttribute.cs:89]: If we use `YAXSerializationOptions`, the setting applies for the whole object graph to serialize. That's what lead to failing unit tests when trying to always include private members from base types.
Agree that `YAXSerializationFields` better were be a flags enum. As it isn't `YAXSerializableTypeAttribute.IncludePrivateMembersFromBaseTypes` became the workaround. What's the way to go:
Bite the bullet and refactor `YAXSerializationFields`?
Leave the workaround?
Move to `YAXSerializationOptions`?

Comment (304NotModified) [reply YAXLib/Attributes/YAXSerializableTypeAttribute.cs:89]: I think we could leave if like this. I don't think there are more reasons to refactor all the stuff. So a bit overkill

**Comment by axunonb** [comment]
> Sorry for the breaking change. I think the tests should be moved to the base classes

Will have a look. There's another issue:

Just tried to merge latest v4.0 into PR #186 and noticed, that - because of new `Serialization` and `Deserialization` classes we need an internal setter for `IRecursionCounter.RecursionCount`. This has few implications that are resolved in a trial now. That could be merged into #186. 
Or do you prefer a separate PR for `IRecursionCounter.RecursionCount`setter?

**Pull request merged**